File: apps/android_native/app/src/main/java/life/devops/passwordmanager/model/EntryTextExport.kt
package life.devops.passwordmanager.model

internal data class EntryTextExportField(
    val id: String,
    val title: String,
)

internal fun VaultEntry.selectedFieldsText(
    fields: List<EntryTextExportField>,
    categoryTemplates: Collection<CategoryTemplate>,
    entries: Collection<VaultEntry>,
): String =
    fields.joinToString("\n") { field ->
        "${field.title.exportLineEscaped()}: ${exportValue(field.id, categoryTemplates, entries).exportLineEscaped()}"
    }

private fun VaultEntry.exportValue(
    fieldId: String,
    categoryTemplates: Collection<CategoryTemplate>,
    entries: Collection<VaultEntry>,
): String {
    when (fieldId) {
        "label" -> return label
        "category" -> return payload.category
        "tags" -> return payload.tags.joinToString(", ")
    }

    when (val entryPayload = payload) {
        is VaultPayload.Credential -> when (fieldId) {
            "credential.username" -> return entryPayload.value.username
            "credential.password" -> return entryPayload.value.password
            "credential.accounts" -> return entryPayload.value.accounts.exportText()
            "credential.token" -> return entryPayload.value.token
            "credential.appId" -> return entryPayload.value.appId
            "credential.accessKey" -> return entryPayload.value.accessKey
            "credential.secretKey" -> return entryPayload.value.secretKey
            "credential.notes" -> return entryPayload.value.notes
        }
        is VaultPayload.Server -> when (fieldId) {
            "server.name" -> return entryPayload.value.name
            "server.ipAddress" -> return entryPayload.value.ipAddress
            "server.port" -> return entryPayload.value.port
            "server.username" -> return entryPayload.value.username
            "server.password" -> return entryPayload.value.password
            "server.accounts" -> return entryPayload.value.accounts.exportText()
            "server.basicConfig" -> return entryPayload.value.basicConfig
            "server.operatingSystem" -> return entryPayload.value.operatingSystem
            "server.location" -> return entryPayload.value.location
            "server.notes" -> return entryPayload.value.notes
        }
        is VaultPayload.Service -> when (fieldId) {
            "service.name" -> return entryPayload.value.name
            "service.connectionAddress" -> return entryPayload.value.connectionAddress
            "service.connectionPort" -> return entryPayload.value.connectionPort
            "service.accountId" -> return referencedEntryText(entryPayload.value.accountId, entries)
            "service.serverIds" -> return entryPayload.value.serverIds
                .joinToString(", ") { referencedEntryText(it, entries) }
            "service.accounts" -> return entryPayload.value.accounts.exportText()
            "service.notes" -> return entryPayload.value.notes
        }
    }

    if (!fieldId.startsWith("custom.")) return ""
    val customFieldId = fieldId.removePrefix("custom.")
    val field = customFields.firstOrNull { it.id == customFieldId } ?: return ""
    resolveFieldReference(
        field = field,
        sourceTemplate = categoryTemplates.firstOrNull {
            it.category.equals(payload.category, ignoreCase = true)
        },
        categoryTemplates = categoryTemplates,
        entries = entries,
    )?.let { return it.exportText() }
    resolveEntryReference(
        field = field,
        template = categoryTemplates.firstOrNull {
            it.category.equals(payload.category, ignoreCase = true)
        },
        entries = entries.toList(),
    )?.let { return it.exportText() }
    return field.value
}

private fun List<ServiceAccount>.exportText(): String =
    joinToString(", ") { account ->
        val note = account.note.takeIf { it.isNotEmpty() }?.let { " - $it" }.orEmpty()
        "${account.username}: ${account.password}$note"
    }

private fun referencedEntryText(id: String?, entries: Collection<VaultEntry>): String {
    if (id.isNullOrBlank()) return ""
    val target = entries.firstOrNull { it.id == id } ?: return "Referenced entry is unavailable."
    if (target.isDeleted) return "Referenced entry was deleted."
    return target.label.ifBlank { "Untitled" }
}

private fun EntryReferenceResolution.exportText(): String =
    when (status) {
        EntryReferenceStatus.EMPTY -> "No entry selected."
        EntryReferenceStatus.RESOLVED -> target?.label?.ifBlank { "Untitled" }
            ?: "Referenced entry is unavailable."
        EntryReferenceStatus.MISSING -> "Referenced entry is unavailable."
        EntryReferenceStatus.DELETED -> "Referenced entry was deleted."
        EntryReferenceStatus.CATEGORY_MISMATCH -> "Referenced entry is outside the target category."
    }

private fun FieldReferenceResolution.exportText(): String =
    when (status) {
        FieldReferenceStatus.EMPTY -> "No entry selected."
        FieldReferenceStatus.INVALID_CONFIGURATION -> "Field reference configuration needs repair."
        FieldReferenceStatus.MISSING -> "Referenced entry is unavailable."
        FieldReferenceStatus.DELETED -> "Referenced entry was deleted."
        FieldReferenceStatus.CATEGORY_MISMATCH -> "Referenced entry is outside the target category."
        FieldReferenceStatus.TARGET_FIELD_MISSING -> "Target field is no longer available."
        FieldReferenceStatus.TARGET_FIELD_UNSUPPORTED -> "Target field is not a text field."
        FieldReferenceStatus.TARGET_FIELD_EMPTY -> "Target field is empty."
        FieldReferenceStatus.RESOLVED -> {
            val targetLabel = targetEntry?.label?.ifBlank { "Untitled" }
                ?: return "Referenced entry is unavailable."
            "$targetLabel: ${targetField?.value.orEmpty()}"
        }
    }

private fun String.exportLineEscaped(): String =
    replace("\r\n", "\\n")
        .replace("\r", "\\n")
        .replace("\n", "\\n")

File: apps/macos_native/Sources/PasswordManagerMacOSApp/Models/EntryTextExport.swift
import Foundation

extension VaultEntry {
    func selectedFieldsText(
        selectedFieldIDs: Set<String>,
        categoryTemplates: [CategoryTemplate],
        entries: [VaultEntry]
    ) -> String {
        exportFields
            .filter { selectedFieldIDs.contains($0.id) }
            .map { field in
                "\(field.title.exportLineEscaped): \(exportValue(for: field.id, categoryTemplates: categoryTemplates, entries: entries).exportLineEscaped)"
            }
            .joined(separator: "\n")
    }

    private func exportValue(
        for fieldID: String,
        categoryTemplates: [CategoryTemplate],
        entries: [VaultEntry]
    ) -> String {
        switch fieldID {
        case "label":
            return label
        case "category":
            return payload.category
        case "tags":
            return payload.tags.joined(separator: ", ")
        default:
            break
        }

        switch payload {
        case .credential(let credential):
            switch fieldID {
            case "credential.username": return credential.username
            case "credential.password": return credential.password
            case "credential.accounts": return credential.accounts.exportText
            case "credential.token": return credential.token
            case "credential.appId": return credential.appId
            case "credential.accessKey": return credential.accessKey
            case "credential.secretKey": return credential.secretKey
            case "credential.notes": return credential.notes
            default: break
            }
        case .server(let server):
            switch fieldID {
            case "server.name": return server.name
            case "server.ipAddress": return server.ipAddress
            case "server.port": return server.port
            case "server.username": return server.username
            case "server.password": return server.password
            case "server.accounts": return server.accounts.exportText
            case "server.accountId": return referencedEntryText(server.accountId, entries: entries)
            case "server.basicConfig": return server.basicConfig
            case "server.operatingSystem": return server.operatingSystem
            case "server.location": return server.location
            case "server.notes": return server.notes
            default: break
            }
        case .service(let service):
            switch fieldID {
            case "service.name": return service.name
            case "service.connectionAddress": return service.connectionAddress
            case "service.connectionPort": return service.connectionPort
            case "service.accountId": return referencedEntryText(service.accountId, entries: entries)
            case "service.serverIds":
                return service.serverIds.map { referencedEntryText($0, entries: entries) }.joined(separator: ", ")
            case "service.accounts": return service.accounts.exportText
            case "service.notes": return service.notes
            default: break
            }
        }

        guard fieldID.hasPrefix("custom.") else { return "" }
        let customFieldID = String(fieldID.dropFirst("custom.".count))
        guard let field = customFields.first(where: { $0.id == customFieldID }) else { return "" }

        if let resolution = resolveFieldReference(
            sourceEntry: self,
            field: field,
            categoryTemplates: categoryTemplates,
            entries: entries
        ) {
            return fieldReferenceExportText(resolution)
        }

        let sourceTemplate = categoryTemplates.first {
            $0.category.caseInsensitiveCompare(payload.category) == .orderedSame
        }
        if let resolution = resolveEntryReference(
            field: field,
            template: sourceTemplate,
            entries: entries
        ) {
            return entryReferenceExportText(resolution)
        }
        return field.value
    }

    private func referencedEntryText(_ id: String?, entries: [VaultEntry]) -> String {
        guard let id, !id.isEmpty else { return "" }
        guard let target = entries.first(where: { $0.id == id }) else {
            return L10n.t("Referenced entry is unavailable.")
        }
        guard !target.isDeleted else { return L10n.t("Referenced entry was deleted.") }
        return target.label.isEmpty ? L10n.t("Untitled") : target.label
    }
}

private extension Array where Element == ServiceAccount {
    var exportText: String {
        map { account in
            let note = account.note.isEmpty ? "" : " - \(account.note)"
            return "\(account.username): \(account.password)\(note)"
        }
        .joined(separator: ", ")
    }
}

private extension String {
    var exportLineEscaped: String {
        replacingOccurrences(of: "\r\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\n")
            .replacingOccurrences(of: "\n", with: "\\n")
    }
}

private func entryReferenceExportText(_ resolution: EntryReferenceResolution) -> String {
    switch resolution.status {
    case .empty:
        return L10n.t("No entry selected.")
    case .resolved:
        guard let target = resolution.target else { return L10n.t("Referenced entry is unavailable.") }
        return target.label.isEmpty ? L10n.t("Untitled") : target.label
    case .missing:
        return L10n.t("Referenced entry is unavailable.")
    case .deleted:
        return L10n.t("Referenced entry was deleted.")
    case .categoryMismatch:
        return L10n.t("Referenced entry is outside the target category.")
    }
}

private func fieldReferenceExportText(_ resolution: FieldReferenceResolution) -> String {
    switch resolution.status {
    case .empty:
        return L10n.t("No entry selected.")
    case .invalidConfiguration:
        return L10n.t("Field reference configuration needs repair.")
    case .missing:
        return L10n.t("Referenced entry is unavailable.")
    case .deleted:
        return L10n.t("Referenced entry was deleted.")
    case .categoryMismatch:
        return L10n.t("Referenced entry is outside the target category.")
    case .targetFieldMissing:
        return L10n.t("Target field is no longer available.")
    case .targetFieldUnsupported:
        return L10n.t("Target field is not a text field.")
    case .targetFieldEmpty:
        return L10n.t("Target field is empty.")
    case .resolved:
        guard let target = resolution.target else { return L10n.t("Referenced entry is unavailable.") }
        let label = target.entryLabel.isEmpty ? L10n.t("Untitled") : target.entryLabel
        return "\(label): \(target.fieldValue)"
    }
}
